crates/wysiwyg/src/composer_model/code_block.rs

```diff
@@ -0,0 +1,523 @@
+        let next_code_block_handle =
+            if let Some(next_sibling) = self.state.dom.next_sibling(&handle) {
+                if next_sibling.kind() == CodeBlock {
+                    Some(next_sibling.handle())
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+        if let Some(next_code_block_handle) = next_code_block_handle {
+            self.move_children_and_delete_parent(
+                &next_code_block_handle,
+                &handle,
+            );
+        }
+
+        let prev_code_block_handle =
+            if let Some(prev_sibling) = self.state.dom.prev_sibling(&handle) {
+                if prev_sibling.kind() == CodeBlock {
+                    Some(prev_sibling.handle())
+                } else {
+                    None
+                }
+            } else {
+                None
+            };
+        if let Some(prev_code_block_handle) = prev_code_block_handle {
+            self.move_children_and_delete_parent(
+                &handle,
+                &prev_code_block_handle,
+            );
+            handle = prev_code_block_handle.clone();
+        }
+
+        self.join_format_nodes_at_level(
+            &handle,
+            handle.raw().len() - 1,
+            &mut DomActionList::default(),
+        );
+        self.join_text_nodes_in_parent(&handle);
+    }
+
+    fn find_nodes_to_wrap_in_block(
+        &self,
+        start: usize,
+        end: usize,
+    ) -> Option<(DomHandle, usize, usize)> {
+        let dom = &self.state.dom;
+        let range = dom.find_range(start, end);
+        let leaves: Vec<&DomLocation> = range.leaves().collect();
+        if leaves.is_empty() {
+            None
+        } else {
+            let first_leaf = leaves.first().unwrap();
+            let last_leaf = leaves.last().unwrap();
+            let iter = dom.iter_from_handle(&last_leaf.node_handle);
+            let rev_iter = dom.iter_from_handle(&first_leaf.node_handle).rev();
+            let mut nodes_to_cover = (
+                HandleWithKind {
+                    handle: first_leaf.node_handle.clone(),
+                    kind: first_leaf.kind.clone(),
+                },
+                HandleWithKind {
+                    handle: last_leaf.node_handle.clone(),
+                    kind: last_leaf.kind.clone(),
+                },
+            );
+            // TODO: check if ancestors of both start and end leaves contains a ListItem instead
+            let selection_contains_several_list_items = range
+                .locations
+                .iter()
+                .filter(|l| l.kind == ListItem)
+                .count()
+                > 1;
+            for node in rev_iter {
+                if !node.is_block_node()
+                    && node.kind() != LineBreak
+                    && (node.kind() != ListItem
+                        || selection_contains_several_list_items)
```

Comment:
Similarly, there is a lot of code that is tightly coupled to lists here. Should code block need to know how lists work?

Reply:
It needs for now, but when I fix the issue with whole lists being included into code blocks instead of only some of their list items I think I might be able to avoid these.